h2o-py/h2o/frame.py

```diff
@@ -1969,22 +1970,22 @@ def as_data_frame(self, use_pandas=True, header=True):
         """
         if can_use_pandas() and use_pandas:
             import pandas
-            if (can_use_datatable()) or (can_use_polars() and can_use_pyarrow()): # can use multi-thread
-                exportFile = tempfile.NamedTemporaryFile(suffix=".h2oframe2Convert.csv", delete=False)
-                try:
-                    exportFile.close()  # needed for Windows
-                    h2o.export_file(self, exportFile.name, force=True)
-                    if can_use_datatable():  # use datatable for multi-thread by default
-                        return self.convert_with_datatable(exportFile.name)
-                    elif can_use_polars() and can_use_pyarrow():  # polar/pyarrow if datatable is not available
-                        return self.convert_with_polars(exportFile.name)
-                finally:
-                    os.unlink(exportFile.name)
+            if use_multi_thread:
+                with local_context(datatable_enabled=True, polars_enabled=True): # turn on multi-thread toolboxes   
```

Comment:
I don't like it - it makes it impossible to choose what library will be used. 

The problem is not just implicit automatic choice but also bugs that are very hard to detect. (We  already know the output differs slightly based on the chosen library.)

Now imagine, in my opinion, a possible scenario:

>Developer develops feature on python 3.10 and since it's in the future when `datatable` supports python 3.10 the developer has both `polars` and `datatable` installed in their environment. But since we choose between the libraries based on the python version, the developer tests everything with `polars` but they don't know they use `polars`, they just use `multi_thread=True`. Everything works so they push it and it fails repeatedly on the CI that uses python 3.9 and H2O then chooses to use `datatable` instead.

I know we mention that multithread uses `polars` or `datatable` in the doc strings but if it's boolean with such an intuitive name I think I would probably skip reading the doc string and since the problem is just on the CI I would not realize it was caused by this soon.

Reply:
Tomas: I have done the following test and note the following:

For Python 3.9 or lower, I used datatable and not polars/pyarrow because the result using polars/pyarrow differ from the single thread converston.

For Python 3.10 and higher, the conversion using polars/pyarrow agrees with the result of single thread conversion.  Datatable does not work with Mac with apple chips.

In the future (I hope near future), we will change the export_file to export parquet which is binary and compressed better and then read in the parquet to pandas directly.  I tried it out yesterday but was running into errors and hence I was just going to push through this simpler version introducing this has_multi_thread flag for the user and they don't need to worry about setting the context as they complained about not wanting to set it.

When we truely finished, we will be using export_file to parquet and then use pandas.read_parquet() command to read in the exported file into pandas directly.  There is no need to choose toolboxes because we will be only using one.

I know this is only temporary.  However, since I don't know when I get around to do it right, I will just do enough to make the customer happy.

What do you think?

Reply:
I don't quite understand what you are trying to say here:

I know we mention that multithread uses polars or datatable in the doc strings but if it's boolean with such an intuitive name I think I would probably skip reading the doc string and since the problem is just on the CI I would not realize it was caused by this soon.

Please rephrase so I can address your concern here.

Reply:
Datatable differs in python 3.9. I noticed it while [working on the jenkins images](https://github.com/h2oai/h2o-3/pull/15964/commits/20a3d2f8987f8c59cdf367f5cdc4ebeb93ec8fd0#r1466276871). 


I modified the failing test so you can check it more easily (the code is below). See how the boolean differs depending on the library used (I added red lines to emphasize what to look at - it's the type of predict column):

![Screenshot 2024-05-30 at 9 38 51](https://github.com/h2oai/h2o-3/assets/61695433/70eb5250-46b8-4448-b7ee-3e2255d41398)

You can try to reproduce it using:

```python
import numpy as np
from IPython.display import Markdown
from h2o.estimators import H2OXGBoostEstimator, H2OIsotonicRegressionEstimator
import os
import tempfile
import h2o
from h2o.utils.threading import local_context
h2o.connect()

def assert_frame_equal(a,b):
    for frame in [a,b]:
        display(Markdown("## Original frame:"))
        display(frame)
        
        display(Markdown("## Single thread:"))
        #display(frame.as_data_frame())
        display(frame.as_data_frame().dtypes)
        
        display(Markdown("## Datatable:"))
        exportFile = tempfile.NamedTemporaryFile(suffix=".h2oframe2Convert.csv", delete=False)
        exportFile.close()  # needed for Windows
        h2o.export_file(frame, exportFile.name, force=True)
        #display(frame.convert_with_datatable(exportFile.name))
        display(frame.convert_with_datatable(exportFile.name).dtypes)
        os.unlink(exportFile.name)
        
        display(Markdown("## Polars:"))
        exportFile = tempfile.NamedTemporaryFile(suffix=".h2oframe2Convert.csv", delete=False)
        exportFile.close()  # needed for Windows
        h2o.export_file(frame, exportFile.name, force=True)
        #display(frame.convert_with_polars(exportFile.name))
        display(frame.convert_with_polars(exportFile.name).dtypes)
        os.unlink(exportFile.name)
        print("-"*80)


df = h2o.import_file(path="http://s3.amazonaws.com/h2o-public-test-data/smalldata/gbm_test/ecology_model.csv")
df["Angaus"] = df["Angaus"].asfactor()
df["Weights"] = h2o.H2OFrame.from_python(abs(np.random.randn(df.nrow, 1)).tolist())[0]

train, calib = df.split_frame(ratios=[.8], destination_frames=["eco_train", "eco_calib"], seed=42)

model = H2OXGBoostEstimator(
    ntrees=100, distribution="bernoulli", min_rows=10, max_depth=5,
    weights_column="Weights",
    calibrate_model=True, calibration_frame=calib, calibration_method="IsotonicRegression"
)
model.train(
    x=list(range(2, train.ncol)),
    y="Angaus", training_frame=train
)

preds_train = model.predict(train)

# Check that calibrated probabilities were appended to the output frame
assert preds_train.col_names == ["predict", "p0", "p1", "cal_p0", "cal_p1"]

preds_calib = model.predict(calib)

isotonic_train = calib[["Angaus", "Weights"]]
isotonic_train = isotonic_train.cbind(preds_calib["p1"])

h2o_iso_reg = H2OIsotonicRegressionEstimator(out_of_bounds="clip")
h2o_iso_reg.train(training_frame=isotonic_train, x="p1", y="Angaus", weights_column="Weights")
print(h2o_iso_reg)

calibrated_p1 = h2o_iso_reg.predict(preds_train)
expected_preds = preds_train[["predict", "p0", "p1"]]
expected_preds["cal_p0"] = 1 - calibrated_p1
expected_preds["cal_p1"] = calibrated_p1

print("-"*80)

assert_frame_equal(expected_preds, preds_train)
```

Reply:
>I know we mention that multithread uses polars or datatable in the doc strings but if it's boolean with such an intuitive name I think I would probably skip reading the doc string and since the problem is just on the CI I would not realize it was caused by this soon.
Please rephrase so I can address your concern here.

Most people don't read documentation if the functionality seems obvious and if it works. When you develop code you test it locally and people usually don't test on the CI changes in single argument if it seems obvious and it works locally -> people will develop some more complicated code and then it'll blow up on the CI. It can be quite hard to find out the root cause of type issues in pandas if you do some pre/post-processing and if you can't reproduce it locally then it's something people can spend hours or days debugging.

So my concern is that this feature can create hard to detect bugs. 

If it works locally and not on the CI which can happen (e.g. developing on M1 mac and CI x86 linux) then I would call that a bad developer experience.

I don't think adding a warning about potential issues with types to the docstrings would help - I think most people would not read it. Maybe having it as a hidden argument (`**kwargs`) would force the user to read the doc string before using it (but I don't like that solution).

If we don't like the solution with `local_context` we could create 2 other functions and make user to use them:

```python
def to_datatable(self):
    "Converts h2oframe to datatable"
def to_polars(self):
    "Converts h2oframe to polars frame"
    
```

And then the user would have a clear idea of what they do:
```python
h2o_frame.to_datatable().to_pandas()

# or

h2o_frame.to_polars().to_pandas()
```

And at the same time it would allow them to use different libraries then pandas.

Reply:
I agree with @tomasfryda that the solution can cause hard-debugged bugs. And I like the suggestion about splitting the functionality into two separate methods. Or I suggest to add another parameter to allow user to use datatable or polars.

Reply:
I will not want to use two solutions.  Datatable is not well maintained and moving forward, I will not want to use it if I can help it.  I am going to check again using only polars/pyarrow and make sure it works for all Python version.

So, for the short term, I will try to only use polars/pyarrow for all Python version.  I will get back to you guys on this.

Reply:
OK, I already agreed with this solution like a year ago.

There was a customer issue `106373` with the polars version. Could you make sure we won't recreate the same bug again?

```python
  File ""/opt/.venv/lib/python3.11/site-packages/src/estimator/predict.py"", line 183, in _predict_datasets
    predictions = predict(model, data, data_path).as_data_frame(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/opt/.venv/lib/python3.11/site-packages/h2o/frame.py"", line 1966, in as_data_frame
    dt_frame = pl.read_csv(fileName)
               ^^^^^^^^^^^^^^^^^^^^^
  File ""/opt/.venv/lib/python3.11/site-packages/polars/io/csv/functions.py"", line 354, in read_csv
    df = pl.DataFrame._read_csv(
         ^^^^^^^^^^^^^^^^^^^^^^^
  File ""/opt/.venv/lib/python3.11/site-packages/polars/dataframe/frame.py"", line 758, in _read_csv
    self._df = PyDataFrame.read_csv(
               ^^^^^^^^^^^^^^^^^^^^^"
2023-08-30 17:11:20.938,"exceptions.ComputeError: Could not parse `""no""` as dtype `bool` at column 'X' (column number 43)."
2023-08-30 17:11:20.938,The current offset in the file is 36707631 bytes.
2023-08-30 17:11:20.938,You might want to try:
2023-08-30 17:11:20.938,"- increasing `infer_schema_length` (e.g. `infer_schema_length=10000`),"
2023-08-30 17:11:20.938,- specifying correct dtype with the `dtypes` argument
2023-08-30 17:11:20.938,"- setting `ignore_errors` to `True`,"
2023-08-30 17:11:20.938,"- adding `""no""` to the `null_values` list."
2023-08-30 17:11:20.938,Closing connection _sid_85dd at exit
2023-08-30 17:11:20.938,H2O session _sid_85dd closed.
```

Reply:
Yes, this is captured in test pyunit_gh_15936_16278_multi_thread_error.py.